### ISSUE TITLE
Gitter vs. Zulip

### ISSUE THREAD

**Issue by stpaultim** [backdrop-ops/backdropcms.org#629]
This is a follow-up to the discussion that has been happening this topic: https://github.com/backdrop-ops/backdropcms.org/issues/587 - Discussion: Alternatives to Gitter

A little over a week ago, we opened dedicated Zulip instance for testing and experimentation. 
https://backdrop.zulipchat.com/

Here are a few other things that have happened:

1. 18 Members of the Backdrop CMS community have created accounts on Zulip and posted hundreds of messages. 
2. @stpaultim Spent some time getting a version of Dragonbot working in Zulip. There are still some issues around hosting for Dragonbot and we have not ported actual data from original Dragonbot (mostly karma points).
3. @laryn experimented with an API for recent activity in Github (Need clarification on exactly how close this is to working). 
4. @BWPanda contacted Zulip and applied for a free hosted Open Source account. Our application was accepted and we now have the benefits of their paid plan. 
5. @Herbdool asked the following question in Gitter, which lead to a discussion with 33 replies posted to the following thread: https://gitter.im/backdrop/backdrop-issues?at=5e17584709534142ca395048

"What needs to still happen for making a decision about whether to stay on Gitter or switch to Zulip?"

In my view, step one to making this decision is probably moving the discussion back to the Github issue queue, which is why I've opened this issue to address this very specific question. How do we proceed with any kind of decision?

I look forward to the discussion. 

---

A PMC Discussion and Vote is in progress on this issue: https://github.com/backdrop-ops/pmc/issues/58

**Comment by stpaultim** [comment]
My personal opinion:

1) I like Zulip better than Gitter. I've never really liked Gitter. 
2) I'm reluctant to push for a change, simply because this is a pretty big change that effects everyone in the community and I'm nervous about the implications of this change, unless there is a strong consensus. It's not at all clear to me that such a consensus exists yet and this issue is NOT that important to me.
3) I understand that there are some legitimate logistical concerns that will need to be dealt with and the potential to loose at least a few people that have created Gitter accounts and might be reluctant to add another chat program to an already overwhelming list of channels of communication. 
4) There is sufficient interest in moving to Zulip that I think we need to make some kind of decision and there should be some clear decision making process that people will feel comfortable with. 
5) If there is significant opposition to any formal move to Zulip, then it might be helpful to get some input from the Project Management Committee on how to handle this?

In short, I would support a move to Zulip if it can be made with some consensus, but I don't feel strongly about it. I think it could be harmful to leave this discussion hanging without any kind of process or formal decision. For me, any decision is better than no decision.

**Comment by docwilmot** [comment]
I'm all in. Zulip is better for me.

**Comment by herbdool** [comment]
I like that Zulip has all features that people like about Gitter plus more. I like that it isn't so buggy (unlike Gitter). I like the topics so I can follow a particular conversation and can find it back later. Much easier if I'm only opening it occasionally.

I like that is open source too and free for an open source project.

**Comment by BWPanda** [comment]
To make it easier to decide/discuss, here are some existing/desired features and how each option addresses that...

### Threaded replies:
The ability to group messages about a single topic/conversation together and browse them separately from the main stream.

**Zulip**
Zulip was built this way from the ground-up and works really well in this regard. There can be multiple streams/channels, and messages within each one are tagged by topic. Topics can be browsed individually or read as part of the main stream. Coming to Zulip after there's been lots of activity is really easy to catch up on what you've missed.

**Gitter**
Gitter now has threaded replies (but wasn't originally built/designed this way), but this feature still seems to be in development as there are some issues people are experiencing (see **Bugs** below). Threads cannot be browsed separately from the main stream, so you have to scroll through it to find them. Coming to Gitter after there's been lots of activity can be confusing to know how/what to catch up on.

I believe the general consensus is that everyone who's tried both options prefers Zulip's implementation of threaded replies.

### Automatic issue linking
The ability to write something like `#123` and have it automatically link to the matching `backdrop-issues` issue queue.

**Zulip**
Zulip has 'linkifiers' that admins can configure which will match a given regular expression and create a link from it. Linkifiers have already been configured in the Backdrop account to allow `#123`, `#pr123` and `org/repo#123` to link to `backdrop-issues` issues, `backdrop` PRs and any other GitHub repo's issues/PRs respectively. These are customisable, so more can be added in future if/as necessary.

**Gitter**
Gitter automatically links `#123` to issues in `backdrop-issues`, but requires the longer syntax (`org/repo#123`) for Backdrop PRs and other repos. There doesn't appear to be a way to customise these.

### GitHub Activity

**Zulip**
Zulip doesn't natively have an activity stream from GitHub, but one can be easily implemented. Some members have already started setting up the necessary webhooks, etc. to test the functionality from their contrib projects. If Zulip is chosen as our official support channel, we can setup an activity stream from the Backdrop issue and PR queues (and again, it's customisable).

**Gitter**
Gitter has a native activity stream (which is where the idea for a Zulip one came from), however it's not customisable and seems to be only for the `backdrop-issues` issue queue (i.e. it doesn't show PRs or activity for contrib projects).

### Forum notifications

**Zulip**
Inspired by what was created for Gitter, discussions have already started about how best to do something similar for Zulip. Based on what else has been possible so far, there should be no problem in setting something up in the regard.

**Gitter**
Gitter has forum notification integration, which is customisable.

### Dragonbot

**Zulip**
The Dragonbot has been setup and configured in Zulip. The only issue currently is that the free hosting we're using requires the bot to sleep a certain number of hours/day, which renders it a bit useless during that time. So paid hosting has been obtained as an interim measure, but other options are being looked into (ultimately this isn't an issue with Zulip, just where to host the bot).

**Gitter**
Gitter has Dragonbot working well. Existing karma points might not be portable to Zulip, but some people have expressed an interest in starting afresh (either when moving to Zulip or as a semi-regular thing in general).

### Existing community

**Zulip**
Some people have created accounts in Zulip already and are using it for announcements, support, testing, etc. Whether this will continue unofficially should Gitter stay the software of choice remains to be seen...

**Gitter**
Obviously this is where the majority of the community meet currently. Probably the biggest reason for staying with Gitter is for this reason - not to lose people. Though suggestions have been made that anyone 'lost' during a migration to Zulip probably aren't active in Gitter/Backdrop anyway (i.e. a lot of users seem to have come over when Backdrop first started but then have disappeared since)...

### Bugs

**Zulip**
In the short testing we've done of Zulip, no one has experienced any major bugs (the one I personally thought I found just turned out to be my intermittent wifi :laughing:). Obviously more testing is needed, but it seems to be a solid platform with great apps, etc.

**Gitter**
Those of us using the Android app have long had problems with it (not refreshing properly when open, issues with notifications, etc.). Now with threaded replies available, we've also discovered that they're not (yet?) visible in the mobile app, so mobile users are basically missing out on those messages altogether...

### Nice-to-haves

There are things that aren't make-or-break either way, but are cool to have if possible:
- Reactions (where you can give a message a thumbs-up, smiley face, etc. without posting a new message in reply): Zulip has this, Gitter doesn't.
- Public archive (a way for 'anonymous' users and search engines to read and index messages): Zulip doesn't have this by default but there's a way to set one up, Gitter has this.

**Comment by BWPanda** [comment]
In case it's not obvious from the way I wrote the above, my vote is for Zulip :slightly_smiling_face: 
Please feel free to mention if anything above should be added, edited or removed.

**Comment by herbdool** [comment]
@BWPanda thanks for the detail! Can you add comparison for signing up? With Gitter people can sign up via Github. And we've found the same is true with Zulip. This was one concern with making sure it was equally easy for people to sign up and login.

Also, I'll note that dragonbot in Gitter has been buggy. It stopped working for me awhile back. And for others it has never worked. We might encounter future issues with dragonbot in Zulip but so far it's been good.

**Comment by BWPanda** [comment]
@herbdool Post updated with sign in info. Re. dragonbot I think those issues are unrelated to Gitter/Zulip, but more like thanks to @stpaultim and co. for fixing all that for us.

**Comment by laryn** [comment]
I prefer Zulip, mostly because it makes it catching up and following (or muting) threads/topics so simple. It does seem like it can match or exceed what Gitter gives us, plus the superior topic/stream functionality.

**Comment by olafgrabienski** [comment]
I'd also vote for Zulip, for similar reasons as already stated by others.

Only drawback in my opinion: the Zulip UI looks more busy and a bit dated. It's however a known issue, there are workarounds, and there seem to be plans to improve the UI, see for instance https://github.com/zulip/zulip-desktop/issues/786.

**Comment by herbdool** [comment]
@olafgrabienski I too hope the UI improves a bit.

One temporary fix I've found is to one, put the user list on the left side (in the settings) and increase the font size in the browser enough so that it becomes two columns.

**Comment by BWPanda** [comment]
Added **Unsubscribe from select notifications** section to my comparison post above.

**Comment by klonos** [comment]
FTR, I prefer Zulip too.

**Comment by Graham-72** [comment]
I prefer Zulip. It makes life easier for following specific topics, and seems to work better when coming back to check for new contributions.

**Comment by quicksketch** [comment]
We have opened a PMC discussion and vote for this question. Even though it's only accessible to PMC members for reference it is https://github.com/backdrop-ops/pmc/issues/58.

**Comment by klonos** [comment]
FTR, the PMC has decided in favor of Zulip, and our https://backdropcms.org/support/resources page has been updated to reflect that (pending official announcement via a blog post).

**Comment by olafgrabienski** [comment]
@klonos In the sidebar of https://backdropcms.org/support/resources there is still the Gitter link, not Zulip.

**Comment by olafgrabienski** [comment]
Two more ideas:

-  Change the headline of https://gitter.im/backdrop/backdrop-issues (*The main chat channel for Backdrop CMS!*).
- Put a sticky message with a link to Zulip at top of the Gitter channel(s), if possible.

**Comment by BWPanda** [comment]
Thanks @olafgrabienski! I've updated the Resources page and the headline of the Gitter channel. Sticky messages aren't yet possible apparently: https://gitlab.com/gitlab-org/gitter/webapp/issues/501

**Comment by BWPanda** [comment]
@stpaultim As our resident chatbot expert, how difficult/time-consuming would it be to add new functionality to Gitter's Dragonbot that:

- Posted a message to Gitter notifying of the move to Zulip every so often, and
- Replied to new messages saying we've moved to Zulip but support *may* still be provided here

There'd obviously need to be some sort of checks so that if there's no activity Dragonbot just doesn't keep posting. Maybe:

- If Dragonbot's last 'Zulip' message was more than a week ago, OR was more than 10 messages ago, post a new one.
- If Dragonbot hasn't already informed X user of the move to Zulip, tell them when they next post.
- Etc.

???

**Comment by BWPanda** [comment]
Cleaning up issue queue.
